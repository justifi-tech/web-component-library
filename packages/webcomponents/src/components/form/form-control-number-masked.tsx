import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';
import IMask, { InputMask } from 'imask';

@Component({
  tag: 'form-control-number-masked'
})
export class NumberInputMasked {
  textInput!: HTMLInputElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() mask: string;
  @Prop() disabled: boolean;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  private imask: InputMask<any> | null = null;

  updateInput(newValue: any) {
    if (this.imask && newValue) {
      this.imask.value = String(newValue);
    }
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.formControlInput.emit({ name: name, value: target.value });
  }

  private initializeIMask() {
    if (!this.textInput) return;

    this.imask = IMask(this.textInput, {
      mask: this.mask,
    });

    this.imask.on('accept', () => {
      const rawValue = this.imask.unmaskedValue;
      this.inputHandler(this.name, rawValue);
    });

    this.textInput.addEventListener('blur', () => this.formControlBlur.emit());
  }

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  componentDidLoad() {
    this.initializeIMask();
    this.updateInput(this.defaultValue);
  }

  disconnectedCallback() {
    this.imask?.destroy();
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <label part="label" class="form-label" htmlFor={this.name}>
            {this.label}
          </label>
          <input
            ref={el => (this.textInput = el as HTMLInputElement)}
            id={this.name}
            name={this.name}
            onInput={(event: any) => this.handleFormControlInput(event)}
            onBlur={() => this.formControlBlur.emit()}
            part={`input ${this.errorText && 'input-invalid'}`}
            class={this.errorText ? 'form-control is-invalid' : 'form-control'}
            type="text"
            disabled={this.disabled}
          />
          <form-control-help-text helpText={this.helpText} />
          <form-control-error-text errorText={this.errorText} />
        </div>
      </Host>
    );
  }
}