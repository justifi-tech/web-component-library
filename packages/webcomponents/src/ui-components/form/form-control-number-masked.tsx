import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
  State,
} from '@stencil/core';
import IMask, { InputMask } from 'imask';
import { FormControlErrorText } from '../../ui-components';
import { input, inputDisabled, inputFocused, inputInvalid, label } from '../../styles/parts';

@Component({
  tag: 'form-control-number-masked'
})
export class NumberInputMasked {
  textInput!: HTMLInputElement;
  private imask: InputMask<any> | null = null;

  @State() isFocused: boolean = false;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() mask: string;
  @Prop() disabled: boolean;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.initializeIMask();
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  disconnectedCallback() {
    this.imask?.destroy();
  }

  componentDidLoad() {
    if (this.textInput) {
      this.initializeIMask();
      this.updateInput(this.defaultValue);
    }
  }

  private initializeIMask = () => {
    if (!this.textInput) return;

    this.imask = IMask(this.textInput, {
      mask: this.mask,
    });

    this.imask.on('accept', () => {
      const rawValue = this.imask?.unmaskedValue;
      if (rawValue !== undefined) {
        this.inputHandler(this.name, rawValue);
      }
    });

    this.textInput.addEventListener('blur', () => this.formControlBlur.emit());
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    const rawValue = this.imask?.unmaskedValue || target.value;
    this.formControlInput.emit({ name: name, value: rawValue });
  }

  updateInput = (newValue: any) => {
    if (this.imask && newValue) {
      this.imask.value = String(newValue);
    }
  }

  private get part() {
    if (this.errorText) {
      return inputInvalid;
    }
    if (this.disabled) {
      return inputDisabled;
    }
    if (this.isFocused) {
      return inputFocused;
    }
    return input;
  }

  render() {
    return (
      <Host>
        <div class="form-group d-flex flex-column">
          <div class="d-flex align-items-start gap-2">
            <label part={label} class="form-label" htmlFor={this.name}>
              {this.label}
            </label>
            <form-control-tooltip helpText={this.helpText} />
          </div>
          <input
            ref={el => (this.textInput = el as HTMLInputElement)}
            id={this.name}
            name={this.name}
            onFocus={() => (this.isFocused = true)}
            onBlur={() => {
              this.isFocused = false;
              this.formControlBlur.emit();
            }}
            onInput={this.handleFormControlInput}
            part={this.part}
            class={this.errorText ? 'form-control is-invalid' : 'form-control'}
            type="text"
            disabled={this.disabled}
          />
          <FormControlErrorText errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
