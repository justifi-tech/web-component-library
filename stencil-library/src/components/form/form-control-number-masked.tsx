import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Watch,
} from '@stencil/core';
import IMask, { InputMask } from 'imask';

@Component({
  tag: 'form-control-number-masked',
  styleUrl: 'form-control-number.scss',
  shadow: true,
})
export class NumberInputMasked {
  @Prop() label: string;
  @Prop() name: any;
  @Prop() error: string;
  @Prop() defaultValue: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() mask: string;
  @Prop() disabled: boolean;

  private imask: InputMask<any> | null = null;

  textInput!: HTMLInputElement;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  componentDidLoad() {
    if (this.textInput && this.mask) {
      this.imask = IMask(this.textInput, {
        mask: this.mask,
      });

      this.imask.on('accept', () => {
        const rawValue = this.imask.unmaskedValue;
        this.inputHandler(this.name, rawValue);
      });

      this.textInput.addEventListener('blur', () => {
        this.formControlBlur.emit();
      });

      this.updateInput(this.defaultValue);
    }
  }

  disconnectedCallback() {
    this.imask?.destroy();
  }

  updateInput(newValue: any) {
    if (this.imask) {
      this.imask.value = String(newValue);
    }
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <input
          ref={el => (this.textInput = el as HTMLInputElement)}
          id={this.name}
          name={this.name}
          onBlur={() => this.formControlBlur.emit()}
          part={`input ${this.error && 'input-invalid'}`}
          class={this.error ? 'form-control is-invalid' : 'form-control'}
          type="text"
          disabled={this.disabled}
        />
        {this.error && <div class="invalid-feedback">{this.error}</div>}
      </Host>
    );
  }
}
