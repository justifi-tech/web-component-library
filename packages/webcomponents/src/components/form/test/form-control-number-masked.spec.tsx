import { newSpecPage } from "@stencil/core/testing";
import { NumberInputMasked } from "../form-control-number-masked";

describe('form-control-number-masked', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [NumberInputMasked],
      html: `<form-control-number-masked label="Age"></form-control-number-masked>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with an error message', async () => {
    const page = await newSpecPage({
      components: [NumberInputMasked],
      html: `<form-control-number-masked label="Age" error-text="Invalid input"></form-control-number-masked>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('handles the disabled state', async () => {
    const page = await newSpecPage({
      components: [NumberInputMasked],
      html: `<form-control-number-masked label="Age" disabled></form-control-number-masked>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('updates the input value when defaultValue changes', async () => {
    const page = await newSpecPage({
      components: [NumberInputMasked],
      html: `<form-control-number-masked mask=${'00-00'} label="Masked Number"></form-control-number-masked>`,
    });
    let input = page.root.querySelector('input');
    expect(input.value).toBe(''); // Default should be empty

    page.root.inputHandler = jest.fn();
    page.root.defaultValue = '1234';
    await page.waitForChanges();

    input = page.root.querySelector('input');
    expect(input.value).toBe('12-34');
  });

  it.only('calls inputHandler on user input', async () => {
    const inputHandlerMock = jest.fn();
    const page = await newSpecPage({
      components: [NumberInputMasked],
      html: `<form-control-number-masked mask=${'00-00'} name='number'></form-control-number-masked>`,
      // template: () => (
      //   <form-control-number-masked mask={PHONE_MASKS.US} name='number' />
      // )
    });

    page.rootInstance.inputHandler = inputHandlerMock;
    await page.waitForChanges();

    const input = page.root.querySelector('input');
    input.value = '1234';
    await input.dispatchEvent(new Event('input', { bubbles: true }));

    expect(inputHandlerMock).toHaveBeenCalledWith('number', '1234');
  });
});