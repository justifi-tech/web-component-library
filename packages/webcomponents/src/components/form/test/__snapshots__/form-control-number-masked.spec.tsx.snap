// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-number-masked handles the disabled state 1`] = `
<form-control-number-masked disabled="" exportparts="label,input,input-invalid" label="Age">
  <div class="d-flex flex-column form-group">
    <label class="form-label" part="label">
      Age
    </label>
    <input class="form-control" disabled="" part="input undefined" type="text">
    <form-control-help-text></form-control-help-text>
    <form-control-error-text></form-control-error-text>
  </div>
</form-control-number-masked>
`;

exports[`form-control-number-masked renders with an error message 1`] = `
<form-control-number-masked error-text="Invalid input" exportparts="label,input,input-invalid" label="Age">
  <div class="d-flex flex-column form-group">
    <label class="form-label" part="label">
      Age
    </label>
    <input class="form-control is-invalid" part="input input-invalid" type="text">
    <form-control-help-text></form-control-help-text>
    <form-control-error-text errortext="Invalid input"></form-control-error-text>
  </div>
</form-control-number-masked>
`;

exports[`form-control-number-masked renders with default props 1`] = `
<form-control-number-masked exportparts="label,input,input-invalid" label="Age">
  <div class="d-flex flex-column form-group">
    <label class="form-label" part="label">
      Age
    </label>
    <input class="form-control" part="input undefined" type="text">
    <form-control-help-text></form-control-help-text>
    <form-control-error-text></form-control-error-text>
  </div>
</form-control-number-masked>
`;
